src/libp2p/Libp2p.Protocols.TLS/TLS.cs

```diff
@@ -0,0 +1,408 @@
+
+
+        _logger?.LogDebug($"New address with PeerID: {newAddressWithPeerId}");
+        context.RemotePeer.Address = newAddressWithPeerId;
+
+
+        _logger?.LogDebug($"Successfully received data from signaling channel. Remote peer address: {remotePeerId}");
+
+        X509Certificate certificate = CertificateHelper.CertificateFromIdentity(_sessionKey, context.LocalPeer.Identity);
+
+        _logger?.LogDebug($"certificate: {certificate}");
+
+        SslServerAuthenticationOptions serverAuthenticationOptions = new()
+        {
+            ApplicationProtocols = [new SslApplicationProtocol("/yamux/1.0.0")],
+            RemoteCertificateValidationCallback = (_, certificate, _, _) => VerifyRemoteCertificate(context.RemotePeer.Address, certificate),
+            ServerCertificate = certificate,
+            ClientCertificateRequired = true,
+        };
+        _logger?.LogTrace("Successfully created client authentication options.");
+        // _logger?.LogDebug("Accepted new TCP client connection from {RemoteEndPoint}", tcpClient.Client.RemoteEndPoint);
+        SslStream sslStream = new(str, false, serverAuthenticationOptions.RemoteCertificateValidationCallback);
+        _logger?.LogTrace("Sslstream initialized.");
+        try
+        {
+            _logger?.LogTrace("Sslstream initialized.1");
+            await sslStream.AuthenticateAsServerAsync(serverAuthenticationOptions);
+            _logger?.LogTrace("Sslstream initialized.2");
+        }
+        catch (Exception ex)
+        {
+            _logger?.LogError("An error occurred during TLS authentication: {Message}", ex.Message);
+            _logger?.LogDebug("Exception details: {StackTrace}", ex.StackTrace);
+            throw;
+        }
+
+        if (sslStream.NegotiatedApplicationProtocol == SslApplicationProtocol.Http2)
+        {
+            _logger?.LogDebug("HTTP/2 protocol negotiated");
+        }
+        else if (sslStream.NegotiatedApplicationProtocol == SslApplicationProtocol.Http11)
+        {
+            _logger?.LogDebug("HTTP/1.1 protocol negotiated");
+        }
+        else if (sslStream.NegotiatedApplicationProtocol == SslApplicationProtocol.Http3)
+        {
+            _logger?.LogDebug("HTTP/3 protocol negotiated");
+        }
+        else
+        {
+            _logger?.LogDebug($"Protocol negotiated!");
+        }
+
+        IChannel upChannel = channelFactory.SubListen(context);
+
+        LastNegotiatedApplicationProtocol = sslStream.NegotiatedApplicationProtocol;
+
+
+        await ExchangeData(sslStream, upChannel, _logger);
+
+    }
+
+
+    private static bool VerifyRemoteCertificate(Multiaddress remotePeerAddress, X509Certificate certificate) =>
+             CertificateHelper.ValidateCertificate(certificate as X509Certificate2, remotePeerAddress.Get<P2P>().ToString());
+
+    public class ChannelStream(IChannel chan, ILogger<TlsProtocol> logger) : Stream
```

Comment:
Could you move it in a separate file? Can be a part of Core project to be reusable by others